File: cli-ruby/table_game.rb
require_relative "robot"

class TableGame
  def initialize(table_size = [5, 5])
    @table_size = table_size

    map_robot_controls
  end

  def execute(commands = [])
    return if commands.empty?

    place_command = commands.shift
    place_robot(place_command[:params])

    commands.each do |command|
      command_name = command[:name]
      @robot_controls[command_name].call
    end
  end

  private

  def place_robot(params)
    @robot = Robot.new(
      initial_position: [params[:x], params[:y]],
      max_area_to_move: @table_size,
      facing: params[:facing]
    )
  end

  def map_robot_controls
    @robot_controls = {
      MOVE: -> { @robot.move },
      LEFT: -> { @robot.left },
      RIGHT: -> { @robot.right },
      REPORT: -> { @robot.current_position_report },
    }
  end
end

File: web-app/src/table_game.js
import Robot from "./robot.js";
import {COMMANDS} from "./constants.js";

class _TableGame {
    constructor(tableSize = [5, 5], onChangeTable) {
        this.tableSize = tableSize;
        this.robot = {};
        this.onChangeTable = onChangeTable || this.defaultOnChangeTable;

        this.mapRobotControls();
    }

    execute(commands) {
        const validsCommands = this.filterValidCommands(commands);
        this.runCommands(validsCommands);
    }

    runCommands(commands) {
        commands?.forEach(command => {
            if (command.name === COMMANDS.place) {
                this.placeRobot(command.params);
            } else {
                this.robotControls[command.name]();
            }
        });
    }

    filterValidCommands(commands) {
        const validCommands = [];

        commands?.forEach(command => {
            const hasPlaceCommand = !(validCommands.length === 0);
            const isPlaceCommand = command.name === COMMANDS.place;
            const isValidPlaceCommand = isPlaceCommand && this.isValidPlaceCommand(command);
            const isValidMovimentCommand = !isPlaceCommand && hasPlaceCommand && this.isValidMovimentCommand(command);

            if (isValidPlaceCommand || isValidMovimentCommand) {
                validCommands.push(command);
            }
        })

        return validCommands;
    }

    isValidPlaceCommand(command) {
        const params = command.params;
        const isValidXCoordinate = params.x >= 0 && params.x < this.tableSize[0];
        const isValidYCoordinate = params.y >= 0 && params.y < this.tableSize[0];

        return isValidXCoordinate && isValidYCoordinate;
    }

    isValidMovimentCommand(command) {
        return this.robotControls[command.name] !== undefined;
    }

    mapRobotControls() {
        this.robotControls = {
            [COMMANDS.move]: () => {
                this.robot.move();
                this.onChangeTable(this.robot.getCurrentPosition());
            },
            [COMMANDS.left]: () => {
                this.robot.left();
                this.onChangeTable(this.robot.getCurrentPosition());
            },
            [COMMANDS.right]: () => {
                this.robot.right();
                this.onChangeTable(this.robot.getCurrentPosition());
            },
            [COMMANDS.report]: () => this.robot.currentPositionReport(),
        };
    }

    placeRobot(params) {
        this.robot = Robot([params.x, params.y], this.tableSize, params.facing);
        this.onChangeTable(this.robot.getCurrentPosition());
    }

    defaultOnChangeTable(position) {}
}

const TableGame = (tableSize, onChangeTable) => {
    const tableGame = new _TableGame(tableSize, onChangeTable);

    return {
        execute: (commands) => tableGame.execute(commands)
    }
}

export default TableGame;